### ISSUE TITLE
feature request: revert or not revert

### ISSUE THREAD

**Issue by joelb123** [mtkennerly/poetry-dynamic-versioning#9]
While it's right there in your docs that any changes made to versions get reverted at exit, I'm having trouble figuring out the use case for this behavior.  I want to be able to tag the revision then run "poetry version" then the tag gets propagated to pyproject.toml and my source code. 

One can enable this by making a "revert" flag in the configuration, which if set false the files do not get reverted.

def deactivate() -> None:
    config = get_config()
    if "revert" in config:
        if not config["revert"]:
            return

**Comment by mtkennerly** [comment]
Hello! Based on this ticket and #10, I think what you're describing is outside the scope of the plugin. With dynamic versioning, the goal is to never commit the version in the repository - pyproject.toml would stay committed with a placeholder like 0.0.0, and the version would only be injected into files during packaging/publishing. That way, in terms of the repository, the tags are the single source of truth for versions.

If you really just want `poetry version` to be aware of your tags, then I would suggest skipping this plugin and using [Dunamai](https://github.com/mtkennerly/dunamai) directly. Here's an example in Bash:

```
poetry version $(dunamai from git)
```

Would that address your use case?

**Comment by joelb123** [comment]
Thanks for your thoughtful answer.   No, what I really want is for the code to have the proper version injected at pre-commit time, ala setuptools_scm.  That way, GitHub and CI tooling get the proper version and publication gets triggered by successful completion of CI testing rather than manually.

Yes, I know it's possible for code to get the version inside pyproject.toml, but that doesn't work once the project is packaged.  

Using your code via a pre-commit hook would also obviate the need for (currently non-existent) poetry hooks and installation into poetry's root environment.  Without the dynamic fanciness, it could be simply installed into the development environment.

**Comment by mtkennerly** [comment]
Perhaps I'm missing something, but if your CI runs `poetry build` to do the packaging, then the plugin would ensure that the packaged code contains the correct version. The dynamic changes are reverted in the working directory, but the packaged artifact contains a static copy of the correct version afterwards, so when an end user installs the artifact, there's nothing dynamic going on with the version at that point. The end result for users is the same as if you had kept and committed the version in pyproject.toml, but without having to.

**Comment by joelb123** [comment]
I'm not saying that I can't make dynamic versioning work, I'm saying that there are many things about it (and your implementation of it) that not to my tastes:

1. I prefer running tests with the real version number rather than the placeholder version.  For some code, version numbering has real side effects beyond version tags (i.e., if version > X.Y.Z do this).
2.  I prefer your package in my developer dependencies rather than in my base environment.
3. I prefer not to run your code run with every python invocation.  The method you have implemented causes sandbox violations when building other code.
4. I prefer version updating via an affirmative  pre-commit hook where failure is obvious.
5. I prefer that the method of version updating not assert requirements elsewhere in the build..
6. I prefer that the method of version updating not assert requirements on other developers who check out the code.  

I suspect that I'm not alone in these preferences.  Your code looks well-written, and with only a few minor tweaks to enable it as a pre-commit hook, I think you'd get a lot more people using it.

**Comment by mtkennerly** [comment]
Those are all fair points. My thoughts:

* Points 2, 3, and 6 should be resolved when Poetry gets real support for plugins, but this is an unfortunate consequence of the current limitations.
* For point 1, I hope this will also be resolved by real plugin support, but it will depend on how that's implemented and when the plugins are activated. Right now, the file changes are made too late when the Python code has already loaded.
* Points 4 and 5 are by design.

If you want a pre-commit hook just to update pyproject.toml, then you could configure a hook with the `poetry version $(dunamai from git)` command I mentioned before. If you also need the other/custom file substitutions that the plugin supports, then what about if I added a CLI mode to activate the plugin's behavior (without reversions) on demand? That CLI could then be invoked as a pre-commit hook. I think that would address your concerns, but without affecting how "plugin mode" behaves (which I see as incompatible with committing the file changes).

**Comment by joelb123** [comment]
A CLI mode would be just perfect.  I would use it in all my poetry-based projects. Thank you for being open to adding it.

**Comment by mtkennerly** [comment]
I just released 0.7.0 with the CLI mode (covered in README under `Command line mode` header). Please give it a try and let me know what you think \:D

**Comment by joelb123** [comment]
It works beautifully, thanks!

There are three minor would-be-nice things:
1. Graceful failure if run on a directory without a pyproject.toml.
2. If a file isn't updated (input and output strings agree), don't write to the file.  As it is now, matching files get touched at each run.
3. Informational output to stderr: new version string, names of files modified.

Here is the section from my .pre-commit-config.yaml that works for me:

- repo: local
  hooks:
  - id: poetry-dynamic-versioning
    name: update version from SCM tags
    entry: poetry-dynamic-versioning
    always_run: true
    exclude: .*
    language: python
    additional_dependencies:
      - poetry-dynamic-versioning

**Comment by mtkennerly** [comment]
Awesome, thanks for the feedback! Those are all good suggestions, so I'll work on them soon (probably tomorrow).

**Comment by mtkennerly** [comment]
These improvements are now available in 0.8.0. Sample output:

```
$ poetry-dynamic-versioning
Version: 0.7.0.post4.dev0+e3d5ce1
Files with substitutions: none
```

```
$ poetry-dynamic-versioning
Version: 0.7.0-post.4+e3d5ce1
Files with substitutions:
  - C:\github\mtkennerly\poetry-dynamic-versioning\tests\project\project\__init__.py
```

```
$ poetry-dynamic-versioning
Error: Unable to find pyproject.toml
```
